File: Data Structures & Algorithms/binary-search/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        start, end = 0, len(nums)-1

        while start<=end:
            distance = int((end-start) / 2)
            half = distance + start
            val = nums[half]

            if val == target:
                return half;
            elif val < target:
                start = half + 1;
            else:
                end = half-1;

        return -1;


        

File: Data Structures & Algorithms/target-sum/submission-2.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        return ways(0, nums, target);
    }

    public int ways(int index, int[] nums, int target) {
        if (target==0 && index==nums.length) {
            return 1;
            //we found a way to sum
        }

        if (index>=nums.length) {
            return 0;
            //we didn't find a way
        }

        //there are two choices we can make, to add the number, or subtract it 

        int num = nums[index];

        return ways(index+1, nums, target-num) + ways(index+1, nums, target+num);

    }
}
